src/duct.py

```diff
@@ -164,62 +166,165 @@ def create_and_parse_args():
+                        break
+                    else:
+                        raise
+                if not data:  # Still open, but no new data to write
+                    break
+                sys.stdout.buffer.write(data)
+                sys.stdout.buffer.flush()
+                self.file.write(
+                    data.decode("utf-8", "replace")
+                )  # Handling decoding errors
+                self.file.flush()
+
+    def close(self):
+        """Close the slave fd and the file when done."""
+        os.close(self.listener_fd)
+        self.file.close()
+
+
+def monitor_process(
+    stdout, stderr, report, process, report_interval, sample_interval, output_prefix
+):
+    while True:
+        if process.poll() is not None:  # the passthrough command has finished
+            if isinstance(stdout, TeeStream):
+                stdout.close()
+            if isinstance(stderr, TeeStream):
+                stderr.close()
+            break
+
+        elapsed_time = time.time() - report.start_time
+        report.collect_sample()
+        if elapsed_time >= (report.number + 1) * report_interval:
+            aggregated = report.aggregate_samples()
+            for pid, pinfo in aggregated.items():
+                with open(
+                    f"{output_prefix}/{pid}_resource_usage.json", "a"
+                ) as resource_statistics_log:
+                    pinfo["elapsed_time"] = elapsed_time
+                    resource_statistics_log.write(json.dumps(aggregated))
+            report.number += 1
+        time.sleep(sample_interval)
+
+
+def prepare_outputs(capture_outputs, outputs, output_prefix):
+    if capture_outputs in ["all", "stdout"] and outputs in ["all", "stdout"]:
+        stdout = TeeStream(f"{output_prefix}/stdout.txt")
+        stdout.start()
+    elif capture_outputs in ["none", "stderr"] and outputs in ["all", "stdout"]:
+        stdout = subprocess.PIPE
+    else:
+        stdout = subprocess.DEVNULL
+
+    if capture_outputs in ["all", "stderr"] and outputs in ["all", "stderr"]:
+        stderr = TeeStream(f"{output_prefix}/stderr.txt")
+        stderr.start()
+    elif capture_outputs in ["none", "stdout"] and outputs in [
+        "all",
+        "stderr",
+    ]:
+        stderr = subprocess.PIPE
+    else:
+        stderr = subprocess.DEVNULL
+    return stdout, stderr
+
+
 def main():
     """A wrapper to execute a command, monitor and log the process details."""
     args = create_and_parse_args()
     os.makedirs(args.output_prefix, exist_ok=True)
-    try:
-        process = subprocess.Popen(
-            [str(args.command)] + args.arguments.copy(),
-            stdout=subprocess.PIPE,
-            stderr=subprocess.PIPE,
-            preexec_fn=os.setsid,
+    stdout, stderr = prepare_outputs(
+        args.capture_outputs, args.outputs, args.output_prefix
+    )
+    process = subprocess.Popen(
+        [str(args.command)] + args.arguments.copy(),
```

Comment:
btw -- why need to copy?

Reply:
Probably not necessary, but this way we don't have to trust that Popen wont modify

Reply:
it shouldn't ... IMHO too paranoid protection, adding then a question to the code reader either it was on purpose for some reason -- after all we do not do `.copy()` every time we pass a list into a function. So IMHO no reason to be special here unless there was a reason